Required marks and target grade

sint18 opened the pull request:
# Implementation of required marks for a target grade

- Implemented a function to show required marks to get a target grade (first class, second class)
- Added a select box and a button in '**Indication**' section
- Added a checkbox called '**Calculate Target Grade**'

## Testing

1. type `npm start`
2. Click '**Clear**' button
3. Put desired module marks for second year
4. **Lock** all the second year modules
5. Leave the final year modules at 0
6. Check '**Calculate Target Grade**'
7. Choose Grade and Click '**Go**'.

## Important

- Do not choose 'third class' if the marks seem to be above third class because it will **_not work_**.
- Do not forget to **_lock_** the modules otherwise it will **_change_**.

## Some Bugs

- For example, first-class is target grade. The required mark shows 70 in 5 modules. But you can slide down up to 68 on module sliders and still shows first class. (This is fine though)
- You can slide all the way down to 40 on one final year module and still shows first class.

sint18 committed:
added open ssl in start script

sint18 committed:
added checkbox and select

sint18 committed:
implemented marks required for grade - draft

sint18 committed:
minor bugs fixed

sint18 committed:
fixed a small bug

wgw0 commented:
Function works well.

Bugs: 1. After locking a module's inputs then pressing `clear` it results in the modules still being locked.
2. Pressing "Go" after the prediction has already been made results in the counters going up by 1 each time.
3. Clear button should only clear the unlocked module inputs, so that the user can do "what-ifs" without having to put in their own grades each time.

In regards to the **Some Bugs** section,

> For example, first-class is target grade. The required mark shows 70 in 5 modules. But you can slide down up to 68 on module sliders and still shows first class. (This is fine though)
You can slide all the way down to 40 on one final year module and still shows first class.

Is this a bug, or is this just because of the way that grades are calculated due to the rules? With grades being ignored due to certain rules.

m0urao reviewed (commented):
I like it and works well but as @wgw0 said, there are some bugs that need improvement. 

Also

> Function works well.
> 
> Bugs: 1. After locking a module's inputs then pressing `clear` it results in the modules still being locked. 2. Pressing "Go" after the prediction has already been made results in the counters going up by 1 each time. 3. Clear button should only clear the unlocked module inputs, so that the user can do "what-ifs" without having to put in their own grades each time.
> 
> In regards to the **Some Bugs** section,
> 
> > For example, first-class is target grade. The required mark shows 70 in 5 modules. But you can slide down up to 68 on module sliders and still shows first class. (This is fine though)
> > You can slide all the way down to 40 on one final year module and still shows first class.
> 
> Is this a bug, or is this just because of the way that grades are calculated due to the rules? With grades being ignored due to certain rules.

Bugs:
1. This doesn't happen to me, so when ``clear`` is pressed modules are unlocked.
2. Yes, needs improvement
3. Yes, needs improvement

New bugs:
1. If we use one of those UP numbers it gives us predicted grades for the final year BUT it doesn't print a GPA.

m0urao commented:
> Function works well.
> 
> Bugs: 1. After locking a module's inputs then pressing `clear` it results in the modules still being locked. 2. Pressing "Go" after the prediction has already been made results in the counters going up by 1 each time. 3. Clear button should only clear the unlocked module inputs, so that the user can do "what-ifs" without having to put in their own grades each time.
> 
> In regards to the **Some Bugs** section,
> 
> > For example, first-class is target grade. The required mark shows 70 in 5 modules. But you can slide down up to 68 on module sliders and still shows first class. (This is fine though)
> > You can slide all the way down to 40 on one final year module and still shows first class.
> 
> Is this a bug, or is this just because of the way that grades are calculated due to the rules? With grades being ignored due to certain rules.

NikBit101 committed:
fixed bug #2

m0urao reviewed (commented):
👍🏻

sint18 committed:
Added a button - clear only unlocked marks

sint18 commented:
## Update


> Bugs: 
>1. After locking a module's inputs then pressing `clear` it results in the modules still being locked. 
>2. Pressing "Go" after the prediction has already been made results in the counters going up by 1 each time. 
>3. Clear button should only clear the unlocked module inputs, so that the user can do "what-ifs" without having to put in their own grades each time.

1.  Haven't fixed yet
2. fixed by @NikBit101 
3. fixed by @sint18 


> > For example, first-class is target grade. The required mark shows 70 in 5 modules. But you can slide down up to 68 on module sliders and still shows first class. (This is fine though)
> > You can slide all the way down to 40 on one final year module and still shows first class.
> 
> Is this a bug, or is this just because of the way that grades are calculated due to the rules? With grades being ignored due to certain rules.

I believe it is just the way that grades are calculated according to the rules.

>New bugs:
>If we use one of those UP numbers it gives us predicted grades for the final year BUT it doesn't print a GPA.

This doesn't happen on my machine.

wgw0 commented:
The current implementation of the fix for bug 2, does not allow for the user to switch between grade targets and repeatedly alter what they are looking for.

Explanation:
If I set all the modules to clear, then select "Second Class, Upper Division" and click "Go". It will set the grades to the required grades, this is what we want. But then if I want to see what the required grades are for "Second Class, Lower Division", or "First Class" I have to refresh the page.

Simply disabling the go button is not the best solution for this as it stops the user having multiple goes without refreshing.

Commit:
Added Go reset when clear button pressed

Commit:
Updated buttons and styling

Commit:
Fixed functionality and ready for styling change

Commit:
Merge pull request #26 from NikBit101/targetgradebuttonfix

Fixed functionality and ready for styling change

wgw0 reviewed (approved):
Approved, merge into master and let's build off this now.
